### PR TITLE
Vectorize Arc.draw.

### PR DESCRIPTION
**Pull request by anntzer**
... by replacing generators into functions working on numpy arrays.

All modified functions are nested and thus not publically accessible.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/patches.py [hunk 1]
```diff
@@ -1619,12 +1617,14 @@ def iter_circle_intersect_on_line(x0, y0, x1, y1):
                 else:
                     sign_dy = 1.0
                 sqrt_discrim = np.sqrt(discrim)
-                for sign in (1., -1.):
-                    x = (D * dy + sign * sign_dy * dx * sqrt_discrim) / dr2
-                    y = (-D * dx + sign * np.abs(dy) * sqrt_discrim) / dr2
-                    yield x, y
+                return np.row_stack(
```
Comment (timhoffm) [inline lib/matplotlib/patches.py]: Don't know. Better to write the 2x2 array out in all components?

Comment (anntzer) [reply lib/matplotlib/patches.py]: would be quite a mouthful, no?

Comment (timhoffm) [reply lib/matplotlib/patches.py]: A loop to define an array always twists my brain :smile: Don't know what's better:

~~~
return np.row_stack(
    [[(D * dy + sign * sign_dy * dx * sqrt_discrim) / dr2,
      (-D * dx + sign * np.abs(dy) * sqrt_discrim) / dr2]
     for sign in [1, -1]])
~~~
or
~~~
return np.array(
    [[D * dy + sign_dy * dx * sqrt_discrim / dr2,
      -D * dx + np.abs(dy) * sqrt_discrim / dr2],
     [D * dy - sign_dy * dx * sqrt_discrim / dr2,
      -D * dx - np.abs(dy) * sqrt_discrim / dr2]])
~~~
or
~~~
return (D * np.array([[dy, -dx],
                      [dy, -dx]])
        + sqrt_discrim / dr2 * np.array([[sign_dy * dx, np.abs(dy)]
                                         [-sign_dy * dx, -np.abs(dy)]]))
~~~

Anyway, I'll approve and you decide which way to go.

**Comment by anntzer** [comment]
Went with the middle suggestion, which is certainly not worse than row_stack :)
Also replaced np.abs by abs.

File: lib/matplotlib/patches.py [hunk 2]
```diff
@@ -1619,12 +1617,15 @@ def iter_circle_intersect_on_line(x0, y0, x1, y1):
...
                 sqrt_discrim = np.sqrt(discrim)
-                for sign in (1., -1.):
-                    x = (D * dy + sign * sign_dy * dx * sqrt_discrim) / dr2
-                    y = (-D * dx + sign * np.abs(dy) * sqrt_discrim) / dr2
-                    yield x, y
+                return np.array(
+                    [[D * dy + sign_dy * dx * sqrt_discrim / dr2,
+                      -D * dx + abs(dy) * sqrt_discrim / dr2],
+                     [D * dy - sign_dy * dx * sqrt_discrim / dr2,
+                      -D * dx - abs(dy) * sqrt_discrim / dr2]])
```
Comment (QuLogic) [inline lib/matplotlib/patches.py]: You lost the parentheses here around some additions.

Comment (anntzer) [reply lib/matplotlib/patches.py]: oops, that's embarassing, fixed

**Comment by QuLogic** [comment]
According to codecov, `line_circle_intersect` is always returning the empty array. I'm not sure how to add a test to trigger the other paths.

**Comment by anntzer** [comment]
Added a test.  Removed specific handling of tangential case because the results are immediately stuffed into a set(), so duplicate angles don't matter (and because I didn't want to write a test for it...).

**Comment by timhoffm** [comment]
Hm, tangential is still a special case. Just removing the special-cased code does not guarantee that the case is handled correctly by the generic code. A test for that would be good nevertheless.

**Comment by anntzer** [comment]
I was going to argue that an additional test for the tangential case was unneeded when I noticed something funny: the "large arc, high accuracy" case (which is what we're talking about here) is actually broken!

Consider
```python
from matplotlib import pyplot as plt, patches as mpatches
ax = plt.figure().add_subplot(111)
# Large arcs that crosses the axes view limits close to x=0.5.
ax.add_patch(mpatches.Arc((-10, 0.5), 21, 21, theta1=-10, theta2=10, edgecolor="b"))
ax.add_patch(mpatches.Arc((-100, 0.5), 201, 201, theta1=-10, theta2=10, edgecolor="r"))
plt.show()
```

The first arc does not trigger the "large arc, high accuracy" case and is drawn, but the second one isn't.  This is a "regression" from #9661 (@QuLogic), although that PR fixed a separate logic error in the code that caused partial arcs to be drawn as full arcs and without using the "large arc, high accuracy" code...  Going back further, it appears that the "large arc, high accuracy" case had been effectively lost since b7479ef (10 years ago).

It's a bit sad, of course, because this entire thing was apparently written (in 7a82ea8 -- 2007) as support for the Mars Phoenix mission (https://matplotlib.org/tutorials/introductory/sample_plots.html?highlight=phoenix#ellipses).  On the other hand, given the amount of time for which this has been broken, and given that no one ever complained about the accuracy loss, I think we can perhaps just drop the accuracy claim and draw the large arc with a standard Bézier approximation.

In fact, I would say that a better approach for drawing arcs may be to add an "ARC" code to Path objects (similar to LINETO/CURVE3/CURVE4) and let the renderers deal with it: at least cairo (https://www.cairographics.org/manual/cairo-Paths.html#cairo-arc), SVG (https://developer.mozilla.org/en-US/docs/Web/SVG/Element/circle) and postscript (http://www.physics.emory.edu/faculty/weeks/graphics/howtops2.html) have primitives for it, and it seems likely that Agg and pdf also have one.

**Comment by QuLogic** [comment]
OK, we have two approvals, but now I'm not sure if you're saying we should fix more things, or merge it anyway?

**Comment by anntzer** [comment]
I would say this doesn't make things worse and improves the testing a bit, so this can be merged unless @timhoffm insists on more tests (let's wait for his opinion), but things are actually pretty broken to start with and probably I'll do bigger surgery later.

**Comment by jklymak** [comment]
I'll merge, and we can add more tests if @timhoffm still thinks we should.

**Comment by timhoffm** [comment]
We'll let's hope the tests are sufficient. Not going to spend time on this for now.

**Pull request merged**